core/mm/vm.c

```diff
@@ -327,7 +327,6 @@ TEE_Result vm_map_pad(struct user_mode_ctx *uctx, vaddr_t *va, size_t len,
 err_rem_reg:
 	TAILQ_REMOVE(&uctx->vm_info.regions, reg, link);
 err_free_reg:
-	mobj_put(reg->mobj);
```

Comment:
sorry, we misunderstood ourselves. This `mobj_put()` is needed to balance `mobj_get()` (line 287) in case the sequecen fails (lines 296, 300, ...).

Want is needed her IMO is:
```patch
 	(...)
 	res = umap_add_region(&uctx->vm_info, reg, pad_begin, pad_end, align);
 	if (res)
-		goto err_free_reg;
+ 		goto err_put_mobj;
	res = alloc_pgt(uctx);
	if (res)
			goto err_free_reg;
 
	(...)
 
 	return TEE_SUCCESS;
 
 err_rem_reg:
 	TAILQ_REMOVE(&uctx->vm_info.regions, reg, link);
+err_put_mobj:
+	mobj_put(reg->mobj);
 err_free_reg:
-	mobj_put(reg->mobj);
 	free(reg);
 	return res;

Reply:
After vm_map_pad done, it will do `mobj_put()`. Does `mobj_put()`(line 157)  balance `mobj_get()` (line 287)?
https://github.com/OP-TEE/optee_os/blob/203ee23d005b2cec2f21b5de334c5a246be32599/core/pta/system.c#L151-L161

Reply:
The picture is:
* `system_map_zi()` gets a refcount on the mobj due to `mobj_with_fobj_alloc()`.
* `system_map_zi()` calls `vm_map_pad()`.
  * `vm_map_pad()` gets a refcount on mobj and maps the object.
  On failure to map, it must release its refcount before returning with an error status.
* `system_map_zi()` releases its refcount on mobj.
   If mobj was successfully mapped, the mapping holds a refcount.
   If mobj failed to be mapped, its recount now reaches 0 and it is destroyed/released.

When the mobj reference will be unmapped, there is a call to `mobj_put(reg->mobj);` to release to refcount held from `vm_map_pad()` and trigger mobj release:
https://github.com/OP-TEE/optee_os/blob/d5ad7ccf818dc8516f2abe9e3262955c91a03627/core/mm/vm.c#L750-L755

Reply:
Ok, I understand the flow. I think I miss some points, thanks